base/some.jl

```diff
@@ -145,3 +145,6 @@ macro something(args...)
     something = GlobalRef(Base, :something)
     return :($something($expr))
 end
+
+==(a::Some, b::Some) = a.value == b.value
```

Comment:
Perhaps `isequal` would be worthwhile as well?

Reply:
Why? `isequal` defaults to using `==`:

```
help?> isequal
search: isequal issetequal allequal isreal issetuid isreadable

  isequal(x, y) -> Bool

[...]

  Implementation
  ≡≡≡≡≡≡≡≡≡≡≡≡≡≡

  The default implementation of isequal calls ==, so a type that does not involve floating-point values generally only needs
  to define ==.
```

Reply:
Because they aren't the same function
```
julia> Some(missing) == Some(missing)
true

julia> Some(missing).value == Some(missing).value
missing

julia> isequal(Some(missing).value, Some(missing).value)
true
```

Reply:
I don't follow what behavior you want me to implement, sorry. Do you mean that there should be a special `isequal` method handling `Some{Missing}`?

Reply:
The behavioral change is intended; the `Some` wrapper should be pretty much agnostic to its wrapped value and should compare like its wrapped value, for the most part. `missing` support is one thing, yes, but that's not something currently handled specially either - please be clear what behavior you're envisioning.

Reply:
Ah sorry, no special behavior, just a method that similarly uses the wrapped value, i.e.
```suggestion
==(a::Some, b::Some) = a.value == b.value
isequal(a::Some, b::Some) = isequal(a.value, b.value)
```
Without that definition, `isequal` would fall back to `==`, which would make `isequal(Some(missing), Some(missing))` return `missing` instead of `true`. That would then be inconsistent with `isequal(missing, missing)`. My point is just that `==` and `isequal` can behave differently since both are user-extensible, so it seems worthwhile to define both to ensure consistency.

Reply:
Ah, gotcha - that makes sense. That's just proper `missing` support then.

Reply:
Actually, this is a bit more tricky than just adding `isequal`, since `Some(missing) == Some(1)` should not be `missing` but `false`, right?

Reply:
Alright, I've pushed a version that only gives `missing` back if one of the arguments actually was a raw `missing`. So `Some(missing)` isn't unwrapped accidentally, at least when comparing to non-missings. 

Might need another method explicitly for `::Some{Missing}, ::Some{Missing}` to prevent an ambiguity..

Reply:
> `Some(missing) == Some(1)` should not be `missing` but `false`, right?

This isn't so obvious to me. I understand that you would like people using `Some` to be protected from a wrapped `missing` to poison the result of `==`, but OTOH that makes the system harder to reason about (and good luck documenting this...). What about `Some(NaN) == Some(NaN)`, should that return `true` because that's more likely to be what people need in this context? Anyway people should use `isequal` like anywhere else if they want to avoid this kind of issue.

Reply:
Why? `==` already states that only `missing` as an argument returns `missing`. It'd be very counterintuitive to get `missing` out of a comparison of objects that aren't themselves `missing`. We don't return `missing` on `Ref(missing) == Ref(1)` either (or any other wrapper in Base), for example - the fact that the object was wrapped is meaningful.

Reply:
At https://github.com/JuliaLang/julia/issues/52484 you complained that the docstring isn't precise enough, and I agree with that. It wouldn't seem unreasonable for users to expect that `Some` behaves likes a collection (which are documented to propagate `missing` with `==` under some conditions), so this should be clarified somewhere. 

> The semantics of `Some` are "This signals to the caller that the object is not `nothing`" and comparing such objects means "are both objects we want to compare the same and not `nothing`?" and the only real answer to that for `Some(missing) == Some(1)` is `false`, not `missing`.

At this level of generality, one could find many other "real" answers that are equally logical.

Reply:
I think it is entirely UNreasonable to expect `Some` to be a "collection". The term is just way too nebulous for that. You can't iterate `Some`. You can't index `Some`. You can only unwrap it with `something`. What makes `Some` a collection here?

Reply:
What makes it somewhat similar to a collection is that with this PR `==(::Some, ::Some)` would call `==` recursively on its elements. If that's not a collection, then nothing in the docstring says how `==` behaves for it, so you need a new docstring.

Reply:
I'm happy to add a new docstring if that's what's needed to justify the behavior of `Some{Missing}` on `==`. I don't necessarily agree that this is needed (after all, the _only requirement_ is that `==` return `missing` if either input is _exactly_ `missing`), but I can add that if you want.